pkg/apk/apk/implementation.go

```diff
@@ -794,6 +814,85 @@ func (a *APK) fetchAlpineKeys(ctx context.Context, alpineVersions []string) erro
 	return nil
 }
 
+// fetchChainguardKeys fetches the public keys for the repositories in the APK database.
+func (a *APK) fetchChainguardKeys(ctx context.Context, repository string) error {
+	ctx, span := otel.Tracer("go-apk").Start(ctx, "fetchChainguardKeys")
+	defer span.End()
+
+	client := a.client
+	discoveryRequest, err := http.NewRequestWithContext(ctx, http.MethodGet, fmt.Sprintf("%s/%s", strings.TrimSuffix(repository, "/"), "apk-configuration"), nil)
+	if err != nil {
+		return err
+	}
+	if err := a.auth.AddAuth(ctx, discoveryRequest); err != nil {
+		return err
+	}
+	discoveryResponse, err := client.Do(discoveryRequest)
+	if err != nil {
+		return fmt.Errorf("failed to perform key discovery: %w", err)
+	}
+	defer discoveryResponse.Body.Close()
+	switch discoveryResponse.StatusCode {
+	case http.StatusNotFound:
+		// This doesn't implement Chainguard-style key discovery.
+		return nil
+
+	case http.StatusOK:
+		// proceed!
+		break
+
+	default:
+		return fmt.Errorf("chainguard key discovery was unsuccessful for repo %s: %v", repository, discoveryResponse.Status)
+	}
+	// Parse our the JWKS URI
+	var discovery struct {
+		JWKSURI string `json:"jwks_uri"`
+	}
+	if err := json.NewDecoder(discoveryResponse.Body).Decode(&discovery); err != nil {
+		return fmt.Errorf("failed to unmarshal discovery payload: %w", err)
+	}
+
+	jwksRequest, err := http.NewRequestWithContext(ctx, http.MethodGet, discovery.JWKSURI, nil)
+	if err != nil {
+		return err
+	}
+	if err := a.auth.AddAuth(ctx, jwksRequest); err != nil {
+		return err
+	}
+	jwksResponse, err := client.Do(jwksRequest)
+	if err != nil {
+		return fmt.Errorf("failed to fetch JWKS: %w", err)
+	}
+	defer jwksResponse.Body.Close()
+	if jwksResponse.StatusCode != http.StatusOK {
+		return fmt.Errorf("failed to fetch JWKS: %v", jwksResponse.Status)
+	}
+
+	jwks := jose.JSONWebKeySet{}
+	if err := json.NewDecoder(jwksResponse.Body).Decode(&jwks); err != nil {
+		return fmt.Errorf("failed to unmarshal JWKS: %w", err)
+	}
+
+	for _, key := range jwks.Keys {
+		filename := filepath.Join(keysDirPath, key.KeyID+".rsa.pub")
+		f, err := a.fs.OpenFile(filename, os.O_CREATE|os.O_WRONLY, 0o644)
+		if err != nil {
+			return fmt.Errorf("failed to open key file %s: %w", filename, err)
+		}
+		defer f.Close()
+
+		pubBytes, err := cryptoutils.MarshalPublicKeyToPEM(key.Key.(*rsa.PublicKey))
```

Comment:
cryptoutils is a kitchen sink. I know we already transitively depend on it via cosign, but I wish we didn't.

Can we just implement this ourselves and save the direct dep we'll have to sever later? It looks pretty simple.

https://github.com/sigstore/sigstore/blob/v1.8.7/pkg/cryptoutils/publickey.go#L76

Reply:
done